### PR TITLE
HIVE-23817. Pushing TopN Key operator PKFK inner joins (amagyar)

### PR DESCRIPTION
**Pull request by zeroflag**
### Pushing the TopNKey operator through PK-FK inner joins.

Example: 

Customer table:

ID (PK) | LAST_NAME
-- | --
1 | Robinson
2 | Jones
3 | Smith
4 | Heisenberg

Order table:

CUSTOMER_ID (FK) | AMOUNT
-- | --
1 | 100
1 | 50
2 | 200
3 | 30
3 | 40

#### Requirements for doing TopN Key pushdown.

* The PRIMARY KEY constraint on Customer.ID that forbids NULL and duplicate values.
* The NOT_NULL constraint on Order.CUSTOMER_ID that forbids NULL values.
* Plus the FOREIGN KEY constraint between Customer.ID and Order.CUSTOMER_ID ensures that exactly one row exists in the Customer table for any given row in the Order table.

In general if the first n of the order by columns are coming from the child table (FK) then we can copy the TopNKey operator with the first n columns and put it before the join. If all columns are coming from the child table we can move the TopNKey operator without keeping the original.

```
SELECT * FROM Customer, Order 
WHERE Customer.ID = Order.CUSTOMER_ID 
ORDER BY Order.AMOUNT, [Order.*], [Customer.*] LIMIT 3;
```

Result:

CUSTOMER.ID (PK) | CUSTOMER.LAST_NAME | ORDER.AMOUNT
-- | -- | --
3 | Smith | 30
3 | Smith | 40
1 | Robinson | 50
1 | Robinson | 100
2 | Jones | 200

Plan

```
Top N Key Operator
          sort order: +
          keys: ORDER.AMOUNT, [ORDER.*]
          top n: 3
  Select Operator (Order)
  [...]
  Join
    [...]
     Top N Key Operator
          sort order: +
          keys: ORDER.AMOUNT, [ORDER.*], [Customer.*]
          top n: 3
 
```

#### Implementation notes

PkFk join information is extracted on the calcite side and it is attached (child table index & name) to the AST as a query hint.
At the physical plan level we make use of this information to decide if we can push through the topn key operator. We also need to get the origins of the columns (in the order by) to see if they're coming from the child table.

### CODE REVIEW

File: parser/src/java/org/apache/hadoop/hive/ql/parse/HiveLexerParent.g [hunk 1]
```diff
@@ -471,6 +471,8 @@ Number
     (Digit)+ ( DOT (Digit)* (Exponent)? | Exponent)?
     ;
 
+PKFK_JOIN: 'PKFK_JOIN';
```
Comment (jcamachor) [inline parser/src/java/org/apache/hadoop/hive/ql/parse/HiveLexerParent.g:474]: Can we prefix with `KW_` and move above with rest of keywords?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/ASTConverter.java [hunk 2]
```diff
@@ -442,6 +450,40 @@ private QueryBlockInfo convertSource(RelNode r) throws CalciteSemanticException
...
+      RelNode nonFkInput = join.getLeft();
+      ast.addChild(pkFkHint(swapSides ? 0 : 1, HiveRelOptUtil.isRowFilteringPlan(mq, nonFkInput)));
+    }
+  }
+
+  private ASTNode pkFkHint(int fkTableIndex, boolean nonFkSideIsFiltered) {
+    ParseDriver parseDriver = new ParseDriver();
+    try {
+      return parseDriver.parseHint(String.format("PKFK_JOIN(%d, %s)",
+              fkTableIndex, nonFkSideIsFiltered ? NON_FK_FILTERED : "notFiltered"));
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/ASTConverter.java:481]: Naming (NON_FK_FILTERED : "notFiltered") is a bit confusing. We can simplify to NON_FK_FILTERED vs NON_FK_NOT_FILTERED? Create String in converter for both (or enum).

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/topnkey/TopNKeyPushdownProcessor.java [hunk 3]
```diff
@@ -255,6 +260,88 @@ private void pushdownThroughLeftOuterJoin(TopNKeyOperator topNKey) throws Semant
     }
   }
 
+  private void pushdownInnerJoin(TopNKeyOperator topNKey, int fkJoinInputIndex, boolean nonFkSideIsFiltered) throws SemanticException {
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/topnkey/TopNKeyPushdownProcessor.java:263]: Can we add a comment to this method? In which cases we can pushdown through inner join?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/topnkey/TopNKeyPushdownProcessor.java [hunk 4]
```diff
@@ -255,6 +260,88 @@ private void pushdownThroughLeftOuterJoin(TopNKeyOperator topNKey) throws Semant
...
+    TopNKeyDesc newTopNKeyDesc = topNKeyDesc.withKeyColumns(prefixLength);
+    newTopNKeyDesc.setKeyColumns(remapColumns(join, fkJoinInput, newTopNKeyDesc.getKeyColumns()));
+    pushdown(copyDown(fkJoinInput, newTopNKeyDesc));
+    if (topNKeyDesc.getKeyColumns().size() == prefixLength) {
+      LOG.debug("Removing {} above {}", topNKey.getName(), join.getName());
+      join.removeChildAndAdoptItsChildren(topNKey);
+    }
+  }
+
+  private int keyColumnPrefixLength(CommonJoinOperator<? extends JoinDesc> join, TopNKeyOperator topNKeyOperator, int expectedTag, List<ExprNodeDesc> keyColumns) {
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/topnkey/TopNKeyPushdownProcessor.java:294]: nit. Some lines are very long. Please split in multiple lines.

Additionally, can we some comments to these utility methods?

File: ql/src/test/queries/clientpositive/topnkey_inner_join.q [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  (3, 40),
+  (1, 100),
+  (1, 50),
+  (3, 30);
+
+set hive.optimize.topnkey=true;
+set hive.optimize.limittranspose=false;
+
+select 'positive: order by columns are coming from child table';
+-- FIXME: explain select * from customer join orders on customer.id = orders.customer_id order by customer.id limit 3;
```
Comment (jcamachor) [inline ql/src/test/queries/clientpositive/topnkey_inner_join.q:27]: I see this example is below. Can FIXME be removed?

File: ql/src/test/queries/clientpositive/topnkey_inner_join.q [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+select 'negative: no PK/FK';
+alter table customer drop constraint pk_customer_id;
+alter table orders drop constraint fk_order_customer_id;
+explain select * from customer join orders on customer.id = orders.customer_id order by customer.id limit 3;
+
+select 'negatie: no RELY';
+alter table customer add constraint pk_customer_id primary key (id) disable novalidate;
+alter table orders add constraint fk_order_customer_id foreign key (customer_id) references customer(id) disable novalidate;
+explain select * from customer join orders on customer.id = orders.customer_id order by customer.id limit 3;
```
Comment (kasakrisz) [inline ql/src/test/queries/clientpositive/topnkey_inner_join.q:49]: It is also worth to execute these queries to catch correctness issues

**Pull request merged**